blockchain/message_pool/src/msgpool.rs

```diff
@@ -0,0 +1,951 @@
+/// This is the mpool provider struct that will let us access and add messages to messagepool.
+/// future
+pub struct MpoolProvider<DB> {
+    cs: ChainStore<DB>,
+}
+
+impl<'db, DB> MpoolProvider<DB>
+where
+    DB: BlockStore,
+{
+    pub fn new(cs: ChainStore<DB>) -> Self
+    where
+        DB: BlockStore,
+    {
+        MpoolProvider { cs }
+    }
+}
+
+impl<DB> Provider for MpoolProvider<DB>
+where
+    DB: BlockStore,
+{
+    fn subscribe_head_changes(&mut self) -> Subscriber<Arc<Tipset>> {
+        self.cs.subscribe()
+    }
+
+    fn get_heaviest_tipset(&mut self) -> Option<Tipset> {
+        let ts = self.cs.heaviest_tipset()?;
+        Some(ts.as_ref().clone())
+    }
+
+    fn put_message(&self, msg: &SignedMessage) -> Result<Cid, Error> {
+        let cid = self
+            .cs
+            .db
+            .put(msg, Blake2b256)
+            .map_err(|err| Error::Other(err.to_string()))?;
+        Ok(cid)
+    }
+
+    fn state_get_actor(&self, addr: &Address, ts: &Tipset) -> Result<ActorState, Error> {
+        let state = StateTree::new_from_root(self.cs.db.as_ref(), ts.parent_state())
+            .map_err(Error::Other)?;
+        let actor = state.get_actor(addr).map_err(Error::Other)?;
+        actor.ok_or_else(|| Error::Other("No actor state".to_owned()))
+    }
+
+    fn messages_for_block(
+        &self,
+        h: &BlockHeader,
+    ) -> Result<(Vec<UnsignedMessage>, Vec<SignedMessage>), Error> {
+        chain::block_messages(self.cs.blockstore(), h).map_err(|err| err.into())
+    }
+
+    fn messages_for_tipset(&self, h: &Tipset) -> Result<Vec<UnsignedMessage>, Error> {
+        self.cs.messages_for_tipset(h).map_err(|err| err.into())
+    }
+
+    fn load_tipset(&self, tsk: &TipsetKeys) -> Result<Tipset, Error> {
+        self.cs.tipset_from_keys(tsk).map_err(|err| err.into())
+    }
+}
+
+/// This is the main MessagePool struct
+pub struct MessagePool<T: 'static> {
+    local_addrs: Vec<Address>,
+    pending: HashMap<Address, MsgSet>,
+    pub cur_tipset: Mutex<Tipset>,
+    api: Arc<Mutex<T>>,
+    pub min_gas_price: BigInt,
+    pub max_tx_pool_size: i64,
+    pub network_name: String,
+    bls_sig_cache: LruCache<Cid, Signature>,
+    sig_val_cache: LruCache<Cid, ()>,
+    local_msgs: HashMap<Vec<u8>, SignedMessage>,
+}
+
+impl<T> MessagePool<T>
+where
+    T: Provider + std::marker::Send,
+{
+    /// Create a new message pool
+    pub fn new(mut api: T, network_name: String) -> Result<Arc<Mutex<MessagePool<T>>>, Error>
+    where
+        T: Provider,
+    {
+        // LruCache sizes have been taken from the lotus implementation
+        let tipset = Mutex::new(
+            api.get_heaviest_tipset()
+                .ok_or_else(|| Error::Other("No ts in api to set as cur_tipset".to_owned()))?,
+        );
+        let bls_sig_cache = LruCache::new(40000);
+        let sig_val_cache = LruCache::new(32000);
+        let api_mutex = Arc::new(Mutex::new(api));
+        let mp = Arc::new(Mutex::new(MessagePool {
```

Comment:
Is it necessary to have the whole MP protected by a mutex? Looks like it should be sufficient that only some fields need to be. Also, If you have mutual exclusion for the whole struct, you automatically have mutual exclusion for its individual fields (i think).

Reply:
Ok, I'll try removing the mutex around messagepool.